### ISSUE TITLE
Diverging implicit expansion for specific configuration of parameters

### ISSUE THREAD

**Issue by luksow** [theiterators/kebs#47]
```scala
import org.scalatest.{FunSuite, Matchers}
import pl.iterators.kebs.json.KebsSpray
import pl.iterators.kebs.tag.meta.tagged
import pl.iterators.kebs.tagged._
import spray.json.{DefaultJsonProtocol, JsonFormat}

package object domain extends Domain {}

@tagged trait Domain {
  trait SomeTag
  type SomeTaggedValue = Int @@ SomeTag
}

object dto {
  import domain._
  case class SomeDto(opt: Option[SomeTaggedValue])
}

class FailingTests extends FunSuite with Matchers with DefaultJsonProtocol with KebsSpray {
  import dto._

  test("fail") {
    implicitly[JsonFormat[Option[SomeDto]]]
  }
}
```

Results in `diverging implicit expansion for type spray.json.JsonFormat[Option[dto.SomeDto]]`. Interestingly, if you delete any `Option` it will work.

My prediction is that implicit priorities are somehow wrong (compiler bug?) but I couldn't decide how/why.

**Comment by luksow** [comment]
```[info] Compiling 1 Scala source to /home/luke/tudum/kebs/tagged-meta/target/scala-2.12/test-classes ...
performing macro expansion new tagged().macroTransform(abstract trait Domain extends scala.AnyRef {
  def $init$() = {
    ()
  };
  abstract trait SomeTag extends scala.AnyRef;
  type SomeTaggedValue = $at$at[Int, SomeTag] 0s
}) at source-/home/luke/tudum/kebs/tagged-meta/src/test/scala/FailingTests.scala,line-9,offset=252
abstract trait Domain extends scala.AnyRef {
  object SomeTag extends scala.AnyRef {
    def <init>() = {
      super.<init>();
      ()
    }; taggedMeta / Test / compileIncremental 0s
    implicit val SomeTaggedValueCaseClass1Rep = new _root_.pl.iterators.kebs.macros.CaseClass1Rep[SomeTaggedValue, Int](((x$1) => SomeTaggedValue.apply(x$1)), identity)
  };=> playJsonSupport / Test / testOnly 0s
  def $init$() = { / Test / testOnly 0s
    () slickSupport / Test / testOnly 0s
  };=> akkaHttpSupport / Test / testOnly 0s
  abstract trait SomeTag extends scala.AnyRef;
  type SomeTaggedValue = $at$at[Int, SomeTag];
  object SomeTaggedValue extends scala.AnyRef {
    def <init>() = {
      super.<init>();
      ()
    };
    def from(arg: Int) = arg.taggedWith[SomeTag];
    def apply(arg: Int) = from(arg)
  }
}
ClassDef(Modifiers(ABSTRACT | DEFAULTPARAM/TRAIT), TypeName("Domain"), List(), Template(List(Select(Ident(scala), TypeName("AnyRef"))), noSelfType, List(ModuleDef(Modifiers(), TermName("SomeTag"), Template(List(Select(Ident(scala), TypeName("AnyRef"))), noSelfType, List(DefDef(Modifiers(), termNames.CONSTRUCTOR, List(), List(List()), TypeTree(), Block(List(pendingSuperCall), Literal(Constant(())))), ValDef(Modifiers(IMPLICIT), TermName("SomeTaggedValueCaseClass1Rep"), TypeTree(), Apply(Select(New(AppliedTypeTree(Select(Select(Select(Select(Select(Ident(termNames.ROOTPKG), TermName("pl")), TermName("iterators")), TermName("kebs")), TermName("macros")), TypeName("CaseClass1Rep")), List(Ident(TypeName("SomeTaggedValue")), Ident(TypeName("Int"))))), termNames.CONSTRUCTOR), List(Function(List(ValDef(Modifiers(PARAM | SYNTHETIC), TermName("x$1"), TypeTree(), EmptyTree)), Apply(Select(Ident(TermName("SomeTaggedValue")), TermName("apply")), List(Ident(TermName("x$1"))))), Ident(TermName("identity")))))))), DefDef(Modifiers(), TermName("$init$"), List(), List(List()), TypeTree(), Block(List(), Literal(Constant(())))), ClassDef(Modifiers(ABSTRACT | INTERFACE | DEFAULTPARAM/TRAIT), TypeName("SomeTag"), List(), Template(List(Select(Ident(scala), TypeName("AnyRef"))), noSelfType, List())), TypeDef(Modifiers(), TypeName("SomeTaggedValue"), List(), AppliedTypeTree(Ident(TypeName("$at$at")), List(Ident(TypeName("Int")), Ident(TypeName("SomeTag"))))), ModuleDef(Modifiers(), TermName("SomeTaggedValue"), Template(List(Select(Ident(scala), TypeName("AnyRef"))), noSelfType, List(DefDef(Modifiers(), termNames.CONSTRUCTOR, List(), List(List()), TypeTree(), Block(List(pendingSuperCall), Literal(Constant(())))), DefDef(Modifiers(), TermName("from"), List(), List(List(ValDef(Modifiers(PARAM), TermName("arg"), Ident(TypeName("Int")), EmptyTree))), TypeTree(), TypeApply(Select(Ident(TermName("arg")), TermName("taggedWith")), List(Ident(TypeName("SomeTag"))))), DefDef(Modifiers(), TermName("apply"), List(), List(List(ValDef(Modifiers(PARAM), TermName("arg"), Ident(TypeName("Int")), EmptyTree))), TypeTree(), Apply(Ident(TermName("from")), List(Ident(TermName("arg")))))))))))
macro expansion is delayed: macros.this.CaseClass1Rep.repFromCaseClass[CC, F1]
performing macro expansion macros.this.CaseClass1Rep.repFromCaseClass[Seq[Option[dto.SomeDto]], F1] at source-/home/luke/tudum/kebs/tagged-meta/src/test/scala/FailingTests.scala,line-23,offset=557
macro expansion has failed: To materialize case class representation, trait Seq must be a case class
[info] /home/luke/tudum/kebs/tagged-meta/src/test/scala/FailingTests.scala:23:15: macros.this.CaseClass1Rep.repFromCaseClass is not a valid implicit value for pl.iterators.kebs.macros.CaseClass1Rep[Seq[Option[dto.SomeDto]],A] because:
[info] hasMatchingSymbol reported error: To materialize case class representation, trait Seq must be a case class
[info]     implicitly[JsonFormat[Seq[Option[SomeDto]]]]
[info]               ^
[info] /home/luke/tudum/kebs/tagged-meta/src/test/scala/FailingTests.scala:23:15: jsonFlatFormat is not a valid implicit value for spray.json.JsonFormat[Seq[Option[dto.SomeDto]]] because:
[info] hasMatchingSymbol reported error: could not find implicit value for parameter rep: pl.iterators.kebs.macros.CaseClass1Rep[Seq[Option[dto.SomeDto]],A]
[info]     implicitly[JsonFormat[Seq[Option[SomeDto]]]]
[info]               ^
macro expansion is delayed: FailingTests.this.jsonFormatN[T]
performing macro expansion FailingTests.this.jsonFormatN[Seq[Option[dto.SomeDto]]] at source-/home/luke/tudum/kebs/tagged-meta/src/test/scala/FailingTests.scala,line-23,offset=557
macro expansion has failed: To materialize RootJsonFormat, trait Seq must be a case class
[info] /home/luke/tudum/kebs/tagged-meta/src/test/scala/FailingTests.scala:23:15: jsonFormatN is not a valid implicit value for spray.json.JsonFormat[Seq[Option[dto.SomeDto]]] because:
[info] hasMatchingSymbol reported error: To materialize RootJsonFormat, trait Seq must be a case class
[info]     implicitly[JsonFormat[Seq[Option[SomeDto]]]]
[info]               ^
macro expansion is delayed: macros.this.CaseClass1Rep.repFromCaseClass[CC, F1]
performing macro expansion macros.this.CaseClass1Rep.repFromCaseClass[Option[dto.SomeDto], F1] at source-/home/luke/tudum/kebs/tagged-meta/src/test/scala/FailingTests.scala,line-23,offset=557
macro expansion has failed: To materialize case class representation, class Option must be a case class
[info] /home/luke/tudum/kebs/tagged-meta/src/test/scala/FailingTests.scala:23:15: macros.this.CaseClass1Rep.repFromCaseClass is not a valid implicit value for pl.iterators.kebs.macros.CaseClass1Rep[Option[dto.SomeDto],A] because:
[info] hasMatchingSymbol reported error: To materialize case class representation, class Option must be a case class
[info]     implicitly[JsonFormat[Seq[Option[SomeDto]]]]
[info]               ^
[info] /home/luke/tudum/kebs/tagged-meta/src/test/scala/FailingTests.scala:23:15: jsonFlatFormat is not a valid implicit value for spray.json.JsonFormat[Option[dto.SomeDto]] because:
[info] hasMatchingSymbol reported error: could not find implicit value for parameter rep: pl.iterators.kebs.macros.CaseClass1Rep[Option[dto.SomeDto],A]
[info]     implicitly[JsonFormat[Seq[Option[SomeDto]]]]
[info]               ^
macro expansion is delayed: FailingTests.this.jsonFormatN[T]
performing macro expansion FailingTests.this.jsonFormatN[Option[dto.SomeDto]] at source-/home/luke/tudum/kebs/tagged-meta/src/test/scala/FailingTests.scala,line-23,offset=557
macro expansion has failed: To materialize RootJsonFormat, class Option must be a case class
[info] /home/luke/tudum/kebs/tagged-meta/src/test/scala/FailingTests.scala:23:15: jsonFormatN is not a valid implicit value for spray.json.JsonFormat[Option[dto.SomeDto]] because:
[info] hasMatchingSymbol reported error: To materialize RootJsonFormat, class Option must be a case class
[info]     implicitly[JsonFormat[Seq[Option[SomeDto]]]]
[info]               ^
macro expansion is delayed: macros.this.CaseClass1Rep.repFromCaseClass[CC, F1]
performing macro expansion macros.this.CaseClass1Rep.repFromCaseClass[dto.SomeDto, F1] at source-/home/luke/tudum/kebs/tagged-meta/src/test/scala/FailingTests.scala,line-23,offset=557
new _root_.pl.iterators.kebs.macros.CaseClass1Rep[dto.SomeDto, Option[domain.SomeTaggedValue]](SomeDto.apply, ((x$1) => x$1.opt))
Apply(Select(New(AppliedTypeTree(Select(Select(Select(Select(Select(Ident(termNames.ROOTPKG), TermName("pl")), TermName("iterators")), TermName("kebs")), TermName("macros")), TypeName("CaseClass1Rep")), List(TypeTree(), TypeTree()))), termNames.CONSTRUCTOR), List(Select(Ident(dto.SomeDto), TermName("apply")), Function(List(ValDef(Modifiers(PARAM | SYNTHETIC), TermName("x$1"), TypeTree(), EmptyTree)), Select(Ident(TermName("x$1")), TermName("opt")))))
macro expansion is delayed: macros.this.CaseClass1Rep.repFromCaseClass[CC, F1]
performing macro expansion macros.this.CaseClass1Rep.repFromCaseClass[Option[domain.SomeTaggedValue], F1] at source-/home/luke/tudum/kebs/tagged-meta/src/test/scala/FailingTests.scala,line-23,offset=557
macro expansion has failed: To materialize case class representation, class Option must be a case class
[info] /home/luke/tudum/kebs/tagged-meta/src/test/scala/FailingTests.scala:23:15: macros.this.CaseClass1Rep.repFromCaseClass is not a valid implicit value for pl.iterators.kebs.macros.CaseClass1Rep[Option[domain.SomeTaggedValue],A] because: => slickSupport / Test / testOnly 0s
[info] hasMatchingSymbol reported error: To materialize case class representation, class Option must be a case class
[info]     implicitly[JsonFormat[Seq[Option[SomeDto]]]]
[info]               ^
[info] /home/luke/tudum/kebs/tagged-meta/src/test/scala/FailingTests.scala:23:15: jsonFlatFormat is not a valid implicit value for spray.json.JsonFormat[Option[domain.SomeTaggedValue]] because:
[info] hasMatchingSymbol reported error: could not find implicit value for parameter rep: pl.iterators.kebs.macros.CaseClass1Rep[Option[domain.SomeTaggedValue],A]
[info]     implicitly[JsonFormat[Seq[Option[SomeDto]]]]
[info]               ^
macro expansion is delayed: FailingTests.this.jsonFormatN[T]
performing macro expansion FailingTests.this.jsonFormatN[Option[domain.SomeTaggedValue]] at source-/home/luke/tudum/kebs/tagged-meta/src/test/scala/FailingTests.scala,line-23,offset=557
macro expansion has failed: To materialize RootJsonFormat, class Option must be a case class
[info] /home/luke/tudum/kebs/tagged-meta/src/test/scala/FailingTests.scala:23:15: jsonFormatN is not a valid implicit value for spray.json.JsonFormat[Option[domain.SomeTaggedValue]] because:
[info] hasMatchingSymbol reported error: To materialize RootJsonFormat, class Option must be a case class
[info]     implicitly[JsonFormat[Seq[Option[SomeDto]]]]
[info]               ^
[info] /home/luke/tudum/kebs/tagged-meta/src/test/scala/FailingTests.scala:23:15: optionFormat is not a valid implicit value for spray.json.JsonFormat[Option[domain.SomeTaggedValue]] because:
[info] hasMatchingSymbol reported error: diverging implicit expansion for type FailingTests.this.JF[domain.SomeTaggedValue]
[info] starting with method jsonFlatFormat in trait KebsSpray
[info]     implicitly[JsonFormat[Seq[Option[SomeDto]]]]
[info]               ^
[info] /home/luke/tudum/kebs/tagged-meta/src/test/scala/FailingTests.scala:23:15: jsonFlatFormat is not a valid implicit value for FailingTests.this.JF[dto.SomeDto] because:
[info] hasMatchingSymbol reported error: diverging implicit expansion for type FailingTests.this.JF[domain.SomeTaggedValue]
[info] starting with method jsonFlatFormat in trait KebsSpray
[info]     implicitly[JsonFormat[Seq[Option[SomeDto]]]]
[info]               ^
macro expansion is delayed: FailingTests.this.jsonFormatN[T]
performing macro expansion FailingTests.this.jsonFormatN[dto.SomeDto] at source-/home/luke/tudum/kebs/tagged-meta/src/test/scala/FailingTests.scala,line-23,offset=557
macro expansion has failed: Flat format preferred
[info] /home/luke/tudum/kebs/tagged-meta/src/test/scala/FailingTests.scala:23:15: jsonFormatN is not a valid implicit value for FailingTests.this.JF[dto.SomeDto] because:
[info] hasMatchingSymbol reported error: Flat format preferred
[info]     implicitly[JsonFormat[Seq[Option[SomeDto]]]]
[info]               ^
[info] /home/luke/tudum/kebs/tagged-meta/src/test/scala/FailingTests.scala:23:15: optionFormat is not a valid implicit value for spray.json.JsonFormat[Option[dto.SomeDto]] because:
[info] hasMatchingSymbol reported error: diverging implicit expansion for type FailingTests.this.JF[domain.SomeTaggedValue]
[info] starting with method jsonFlatFormat in trait KebsSpray
[info]     implicitly[JsonFormat[Seq[Option[SomeDto]]]]
[info]               ^
[info] /home/luke/tudum/kebs/tagged-meta/src/test/scala/FailingTests.scala:23:15: seqFormat is not a valid implicit value for spray.json.JsonFormat[Seq[Option[dto.SomeDto]]] because:
[info] hasMatchingSymbol reported error: diverging implicit expansion for type FailingTests.this.JF[domain.SomeTaggedValue]
[info] starting with method jsonFlatFormat in trait KebsSpray
[info]     implicitly[JsonFormat[Seq[Option[SomeDto]]]]
[info]               ^
macro expansion is delayed: macros.this.CaseClass1Rep.repFromCaseClass[CC, F1]
performing macro expansion macros.this.CaseClass1Rep.repFromCaseClass[Seq[Option[dto.SomeDto]], F1] at source-/home/luke/tudum/kebs/tagged-meta/src/test/scala/FailingTests.scala,line-23,offset=557
macro expansion has failed: To materialize case class representation, trait Seq must be a case class
[info] /home/luke/tudum/kebs/tagged-meta/src/test/scala/FailingTests.scala:23:15: macros.this.CaseClass1Rep.repFromCaseClass is not a valid implicit value for pl.iterators.kebs.macros.CaseClass1Rep[Seq[Option[dto.SomeDto]],A] because:
[info] hasMatchingSymbol reported error: To materialize case class representation, trait Seq must be a case class
[info]     implicitly[JsonFormat[Seq[Option[SomeDto]]]]
[info]               ^
[info] /home/luke/tudum/kebs/tagged-meta/src/test/scala/FailingTests.scala:23:15: jsonFlatFormat is not a valid implicit value for spray.json.JsonFormat[Seq[Option[dto.SomeDto]]] because:
[info] hasMatchingSymbol reported error: could not find implicit value for parameter rep: pl.iterators.kebs.macros.CaseClass1Rep[Seq[Option[dto.SomeDto]],A]
[info]     implicitly[JsonFormat[Seq[Option[SomeDto]]]]
[info]               ^
macro expansion is delayed: FailingTests.this.jsonFormatN[T]
performing macro expansion FailingTests.this.jsonFormatN[Seq[Option[dto.SomeDto]]] at source-/home/luke/tudum/kebs/tagged-meta/src/test/scala/FailingTests.scala,line-23,offset=557
macro expansion has failed: To materialize RootJsonFormat, trait Seq must be a case class
[info] /home/luke/tudum/kebs/tagged-meta/src/test/scala/FailingTests.scala:23:15: jsonFormatN is not a valid implicit value for spray.json.JsonFormat[Seq[Option[dto.SomeDto]]] because:
[info] hasMatchingSymbol reported error: To materialize RootJsonFormat, trait Seq must be a case class
[info]     implicitly[JsonFormat[Seq[Option[SomeDto]]]]
[info]               ^
macro expansion is delayed: macros.this.CaseClass1Rep.repFromCaseClass[CC, F1]
performing macro expansion macros.this.CaseClass1Rep.repFromCaseClass[Option[dto.SomeDto], F1] at source-/home/luke/tudum/kebs/tagged-meta/src/test/scala/FailingTests.scala,line-23,offset=557
macro expansion has failed: To materialize case class representation, class Option must be a case class
[info] /home/luke/tudum/kebs/tagged-meta/src/test/scala/FailingTests.scala:23:15: macros.this.CaseClass1Rep.repFromCaseClass is not a valid implicit value for pl.iterators.kebs.macros.CaseClass1Rep[Option[dto.SomeDto],A] because:
[info] hasMatchingSymbol reported error: To materialize case class representation, class Option must be a case class
[info]     implicitly[JsonFormat[Seq[Option[SomeDto]]]]
[info]               ^
[info] /home/luke/tudum/kebs/tagged-meta/src/test/scala/FailingTests.scala:23:15: jsonFlatFormat is not a valid implicit value for spray.json.JsonFormat[Option[dto.SomeDto]] because:
[info] hasMatchingSymbol reported error: could not find implicit value for parameter rep: pl.iterators.kebs.macros.CaseClass1Rep[Option[dto.SomeDto],A]
[info]     implicitly[JsonFormat[Seq[Option[SomeDto]]]]
[info]               ^
macro expansion is delayed: FailingTests.this.jsonFormatN[T]
performing macro expansion FailingTests.this.jsonFormatN[Option[dto.SomeDto]] at source-/home/luke/tudum/kebs/tagged-meta/src/test/scala/FailingTests.scala,line-23,offset=557
macro expansion has failed: To materialize RootJsonFormat, class Option must be a case class
[info] /home/luke/tudum/kebs/tagged-meta/src/test/scala/FailingTests.scala:23:15: jsonFormatN is not a valid implicit value for spray.json.JsonFormat[Option[dto.SomeDto]] because:
[info] hasMatchingSymbol reported error: To materialize RootJsonFormat, class Option must be a case class
[info]     implicitly[JsonFormat[Seq[Option[SomeDto]]]]
[info]               ^
macro expansion is delayed: macros.this.CaseClass1Rep.repFromCaseClass[CC, F1]
performing macro expansion macros.this.CaseClass1Rep.repFromCaseClass[dto.SomeDto, F1] at source-/home/luke/tudum/kebs/tagged-meta/src/test/scala/FailingTests.scala,line-23,offset=557
new _root_.pl.iterators.kebs.macros.CaseClass1Rep[dto.SomeDto, Option[domain.SomeTaggedValue]](SomeDto.apply, ((x$2) => x$2.opt))
Apply(Select(New(AppliedTypeTree(Select(Select(Select(Select(Select(Ident(termNames.ROOTPKG), TermName("pl")), TermName("iterators")), TermName("kebs")), TermName("macros")), TypeName("CaseClass1Rep")), List(TypeTree(), TypeTree()))), termNames.CONSTRUCTOR), List(Select(Ident(dto.SomeDto), TermName("apply")), Function(List(ValDef(Modifiers(PARAM | SYNTHETIC), TermName("x$2"), TypeTree(), EmptyTree)), Select(Ident(TermName("x$2")), TermName("opt")))))
macro expansion is delayed: macros.this.CaseClass1Rep.repFromCaseClass[CC, F1]
performing macro expansion macros.this.CaseClass1Rep.repFromCaseClass[Option[domain.SomeTaggedValue], F1] at source-/home/luke/tudum/kebs/tagged-meta/src/test/scala/FailingTests.scala,line-23,offset=557
macro expansion has failed: To materialize case class representation, class Option must be a case class
[info] /home/luke/tudum/kebs/tagged-meta/src/test/scala/FailingTests.scala:23:15: macros.this.CaseClass1Rep.repFromCaseClass is not a valid implicit value for pl.iterators.kebs.macros.CaseClass1Rep[Option[domain.SomeTaggedValue],A] because: => slickSupport / Test / testOnly 0s
[info] hasMatchingSymbol reported error: To materialize case class representation, class Option must be a case class
[info]     implicitly[JsonFormat[Seq[Option[SomeDto]]]]
[info]               ^
[info] /home/luke/tudum/kebs/tagged-meta/src/test/scala/FailingTests.scala:23:15: jsonFlatFormat is not a valid implicit value for spray.json.JsonFormat[Option[domain.SomeTaggedValue]] because:
[info] hasMatchingSymbol reported error: could not find implicit value for parameter rep: pl.iterators.kebs.macros.CaseClass1Rep[Option[domain.SomeTaggedValue],A]
[info]     implicitly[JsonFormat[Seq[Option[SomeDto]]]]
[info]               ^
macro expansion is delayed: FailingTests.this.jsonFormatN[T]
performing macro expansion FailingTests.this.jsonFormatN[Option[domain.SomeTaggedValue]] at source-/home/luke/tudum/kebs/tagged-meta/src/test/scala/FailingTests.scala,line-23,offset=557
macro expansion has failed: To materialize RootJsonFormat, class Option must be a case class
[info] /home/luke/tudum/kebs/tagged-meta/src/test/scala/FailingTests.scala:23:15: jsonFormatN is not a valid implicit value for spray.json.JsonFormat[Option[domain.SomeTaggedValue]] because:
[info] hasMatchingSymbol reported error: To materialize RootJsonFormat, class Option must be a case class
[info]     implicitly[JsonFormat[Seq[Option[SomeDto]]]]
[info]               ^
[info] /home/luke/tudum/kebs/tagged-meta/src/test/scala/FailingTests.scala:23:15: optionFormat is not a valid implicit value for spray.json.JsonFormat[Option[domain.SomeTaggedValue]] because:
[info] hasMatchingSymbol reported error: diverging implicit expansion for type FailingTests.this.JF[domain.SomeTaggedValue]
[info] starting with method jsonFlatFormat in trait KebsSpray
[info]     implicitly[JsonFormat[Seq[Option[SomeDto]]]]
[info]               ^
[info] /home/luke/tudum/kebs/tagged-meta/src/test/scala/FailingTests.scala:23:15: jsonFlatFormat is not a valid implicit value for FailingTests.this.JF[dto.SomeDto] because:
[info] hasMatchingSymbol reported error: diverging implicit expansion for type FailingTests.this.JF[domain.SomeTaggedValue]
[info] starting with method jsonFlatFormat in trait KebsSpray
[info]     implicitly[JsonFormat[Seq[Option[SomeDto]]]]
[info]               ^
macro expansion is delayed: FailingTests.this.jsonFormatN[T]
performing macro expansion FailingTests.this.jsonFormatN[dto.SomeDto] at source-/home/luke/tudum/kebs/tagged-meta/src/test/scala/FailingTests.scala,line-23,offset=557
macro expansion has failed: Flat format preferred
[info] /home/luke/tudum/kebs/tagged-meta/src/test/scala/FailingTests.scala:23:15: jsonFormatN is not a valid implicit value for FailingTests.this.JF[dto.SomeDto] because:
[info] hasMatchingSymbol reported error: Flat format preferred
[info]     implicitly[JsonFormat[Seq[Option[SomeDto]]]]
[info]               ^
[info] /home/luke/tudum/kebs/tagged-meta/src/test/scala/FailingTests.scala:23:15: optionFormat is not a valid implicit value for spray.json.JsonFormat[Option[dto.SomeDto]] because:
[info] hasMatchingSymbol reported error: diverging implicit expansion for type FailingTests.this.JF[domain.SomeTaggedValue]
[info] starting with method jsonFlatFormat in trait KebsSpray
[info]     implicitly[JsonFormat[Seq[Option[SomeDto]]]]
[info]               ^
[info] /home/luke/tudum/kebs/tagged-meta/src/test/scala/FailingTests.scala:23:15: seqFormat is not a valid implicit value for spray.json.JsonFormat[Seq[Option[dto.SomeDto]]] because:
[info] hasMatchingSymbol reported error: diverging implicit expansion for type FailingTests.this.JF[domain.SomeTaggedValue]
[info] starting with method jsonFlatFormat in trait KebsSpray
[info]     implicitly[JsonFormat[Seq[Option[SomeDto]]]]
[info]               ^
[error] /home/luke/tudum/kebs/tagged-meta/src/test/scala/FailingTests.scala:23:15: diverging implicit expansion for type spray.json.JsonFormat[Seq[Option[dto.SomeDto]]]
[error] starting with method jsonFlatFormat in trait KebsSpray
[error]     implicitly[JsonFormat[Seq[Option[SomeDto]]]]
[error]               ^
performing macro expansion source.this.Position.here at source-/home/luke/tudum/kebs/tagged-meta/src/test/scala/FailingTests.scala,line-22,offset=541
_root_.org.scalactic.source.Position.apply("FailingTests.scala", "Please set the environment variable SCALACTIC_FILL_FILE_PATHNAMES to yes at compile time to enable this feature.", 22)
Apply(Select(Select(Select(Select(Select(Ident(termNames.ROOTPKG), TermName("org")), TermName("scalactic")), TermName("source")), TermName("Position")), TermName("apply")), List(Literal(Constant("FailingTests.scala")), Literal(Constant("Please set the environment variable SCALACTIC_FILL_FILE_PATHNAMES to yes at compile time to enable this feature.")), Literal(Constant(22))))
[error] one error found
```
Error log with some extra logging.

**Comment by zofja** [comment]
This test passes successfully:
```
import org.scalatest.funsuite.AnyFunSuite
import org.scalatest.matchers.should.Matchers
import pl.iterators.kebs.json.KebsSpray
import pl.iterators.kebs.tag.meta.tagged
import pl.iterators.kebs.tagged._
import spray.json.{DefaultJsonProtocol, JsonFormat}

package object domain extends Domain {}

@tagged trait Domain {
  trait SomeTag
  type SomeTaggedValue = Int @@ SomeTag
}

object dto {
  import domain._
  case class SomeDto(opt: Option[SomeTaggedValue])
}

class FailingTests extends AnyFunSuite with Matchers with DefaultJsonProtocol with KebsSpray {
  import dto._

  test("maybe successful after all") {
    """implicitly[JsonFormat[Option[SomeDto]]]""".stripMargin should compile
  }
}
```
using following `build.sbt`:
```
name := "kebs-issue-47"
version := "0.1"
scalaVersion := "2.13.6"
scalacOptions := Seq(
  "-target:jvm-1.8",
  "-unchecked",
  "-deprecation",
  "-feature",
  "-encoding",
  "utf8",
  "-Ymacro-annotations",
)

val kebsV = "1.9.2-SNAPSHOT"
val scalatestV = "3.2.9"
val sprayJsonV = "1.3.6"

libraryDependencies ++= Seq(
  "org.scalatest" %% "scalatest" % scalatestV,
  "io.spray" %% "spray-json" % sprayJsonV,
  "pl.iterators" %% "kebs-spray-json" % kebsV,
  "pl.iterators" %% "kebs-tagged-meta" % kebsV
)
```
sbt 1.4.7 and Java 1.8.0_292.

**Comment by luksow** [comment]
@zofja oh that's cool. CC: @kristerr who bumped into it several times.

@zofja could you add a test to the test suite? And then maybe in which Scala version it was fixed? Because I assume it was some kind of compiler bug that got fixed.

**Comment by zofja** [comment]
Seems like it was fixed in Scala 2.13.1. Test passes with following `build.sbt`:
```scala
name := "kebs-issue-47"
version := "0.1"
scalaVersion := "2.13.1"
scalacOptions := Seq(
  "-target:jvm-1.8",
  "-unchecked",
  "-deprecation",
  "-feature",
  "-encoding",
  "utf8",
  "-Ymacro-annotations",
)

val kebsV = "1.9.1"
val scalatestV = "3.2.9"
val sprayJsonV = "1.3.6"

libraryDependencies ++= Seq(
  "org.scalatest" %% "scalatest" % scalatestV,
  "io.spray" %% "spray-json" % sprayJsonV,
  "pl.iterators" %% "kebs-spray-json" % kebsV,
  "pl.iterators" %% "kebs-tagged-meta" % kebsV
)
```
and fails with the following one:
```scala
name := "kebs-issue-47"
version := "0.1"
scalaVersion := "2.12.13"
scalacOptions := Seq(
  "-target:jvm-1.8",
  "-unchecked",
  "-deprecation",
  "-feature",
  "-encoding",
  "utf8"
)

addCompilerPlugin("org.scalamacros" % "paradise" % "2.1.1" cross CrossVersion.full)

val kebsV = "1.9.1"
val scalatestV = "3.2.9"
val sprayJsonV = "1.3.6"

libraryDependencies ++= Seq(
  "org.scalatest" %% "scalatest" % scalatestV,
  "io.spray" %% "spray-json" % sprayJsonV,
  "pl.iterators" %% "kebs-spray-json" % kebsV,
  "pl.iterators" %% "kebs-tagged-meta" % kebsV
)
```
with the expected error: `diverging implicit expansion for type spray.json.JsonFormat[Option[dto.SomeDto]]`

**Comment by luksow** [comment]
Very cool, thanks!
